modules/340-monitoring-kubernetes/images/oom-kills-exporter/src/main.go

```diff
@@ -0,0 +1,129 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"errors"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/kmsg"
+	"k8s.io/node-problem-detector/pkg/systemlogmonitor/logwatchers/types"
+)
+
+var (
+	local = prometheus.NewRegistry()
+)
+
+func checkMetricExistenceByLabels(metricName string, labels map[string]string, r *prometheus.Registry) bool {
+	mfs, err := r.Gather()
+	if err != nil {
+		log.Println("Error gathering metrics:", err)
+		return false
+	}
+
+	for _, mf := range mfs {
+		if mf.GetName() == metricName {
+			for _, metric := range mf.Metric {
+				labelMap := make(map[string]string)
+				for _, label := range metric.Label {
+					labelMap[*label.Name] = *label.Value
+				}
+
+				match := true
+				for key, value := range labels {
+					if labelMap[key] != value {
+						match = false
+						break
+					}
+				}
+
+				if match {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func getContainerIDFromLog(line string) (string, error) {
+	var taskMemcg string
+	if matches := regexp.MustCompile(`^oom-kill:(.+)`).FindStringSubmatch(line); matches != nil {
+		log.Print(line)
+		for _, word := range strings.Split(matches[0], ",") {
+			if strings.Contains(word, "task_memcg") {
+				if idx := strings.Index(word, "="); idx != -1 {
+					taskMemcg = word[idx+1:]
+				}
+			}
+		}
+		return taskMemcg, nil
+	}
+	return "", errors.New("Don't oom-kill log")
+}
+
+func main() {
+	handler := promhttp.HandlerFor(
+		local,
+		promhttp.HandlerOpts{
+			EnableOpenMetrics: false,
+		})
+	klogOomkill := prometheus.NewCounterVec(
+		prometheus.CounterOpts{Name: "klog_oomkill"},
+		[]string{"task_memcg"},
+	)
+	local.MustRegister(klogOomkill)
+
+	go func() {
+		log.Print("Starting prometheus metrics")
+		http.HandleFunc("/ready", func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write([]byte("ok"))
+		})
+		http.Handle("/metrics", handler)
+		log.Fatal(http.ListenAndServe("127.0.0.1:4205", nil))
+	}()
+
+	kmsgWatcher := kmsg.NewKmsgWatcher(types.WatcherConfig{Plugin: "kmsg", Lookback: "240h"})
+	logCh, err := kmsgWatcher.Watch()
+	if err != nil {
+		log.Fatal("Could not create log watcher")
+	}
+
+	for item := range logCh {
+		if taskMemcg, err := getContainerIDFromLog(item.Message); err == nil {
+			labels := map[string]string{
+				"task_memcg": taskMemcg,
+			}
+			sleep := 0 * time.Second
+			if !checkMetricExistenceByLabels("klog_oomkill", labels, local) {
+				if _, err := klogOomkill.GetMetricWith(labels); err != nil {
+					log.Fatal("Could not create metrics")
+				}
+				sleep = 31 * time.Second
```

Comment:
what is magic number?

Reply:
This is the default prometeus scrape interval we have in the platform. I have not found access to the variable to get its value directly into the exporter.

Reply:
during a conversation with Alexey, we decided to use `prometheusScrapeInterval' from global values for this variable

Reply:
Modified to use a variable.